### PR TITLE
[DEV-470] - Implement updated permissions

### PR DESCRIPTION
**Pull request by grp06**
## Describe your changes
- the RAFT holder can set permissions for who is allowed to create new badge specs
- it can either be 1. only raft holder 2. any badge holder 3. specific badge holders

- I went back to a single function to create a spec `createSpec()` ... inside this function we check the "spec creation permissions" and proceed accordingly. If the permissions say that only the raft holder can create specs, we have a require statement checking that, etc...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by rahulrumalla** [comment]
Can you provide some context on what made you consider importing `ERC721EnumberableUpgradeable` into `Badges.sol`?

File: src/Badges.sol [hunk 1]
```diff
@@ -32,6 +32,7 @@ contract Badges is
   mapping(address => uint256) private balances;
 
   ISpecDataHolder private specDataHolder;
+  mapping(uint256 => uint256) raftToBadgeSpecCreationPermissions;
```
Comment (rahulrumalla) [inline src/Badges.sol:35]: Consider using an `enum` instead of a `uint256`

File: src/Badges.sol [hunk 2]
```diff
@@ -65,6 +66,16 @@ contract Badges is
       super.supportsInterface(_interfaceId);
   }
 
+  // permissionInt:
+  // 0: only raft owner
+  // 1: all badge holders associated with that RAFT
+  // 2: only specific badge holders
+  function setBadgeSpecCreationPermissions(uint256 raftTokenId, uint256 permissionInt) external {
```
Comment (rahulrumalla) [inline src/Badges.sol:73]: There's no authorizations on this function. Would it be `onlyOwner` ?

File: src/Badges.sol [hunk 3]
```diff
@@ -101,12 +112,20 @@ contract Badges is
...
-    require(!specDataHolder.isSpecRegistered(_specUri), "createSpec: spec already registered");
-
+  function createSpecAsRaftHolder(string memory _specUri, uint256 _raftTokenId) external virtual {
+    require(specDataHolder.getRaftOwner(_raftTokenId) == msg.sender, "createSpecAsRaftHolder: unauthorized");
+    require(!specDataHolder.isSpecRegistered(_specUri), "createSpecAsRaftHolder: spec already registered");
     specDataHolder.setSpecToRaft(_specUri, _raftTokenId);
+    emit SpecCreated(msg.sender, _specUri, _raftTokenId, specDataHolder.getRaftAddress());
+  }
 
+  function createSpecAsHolderOfAnyBadge(
```
Comment (rahulrumalla) [inline src/Badges.sol:122]:
```suggestion
  function createSpecAsBadgeHolder(
```

Keeping it consistent with the other function name

**Comment by grp06** [comment]
Note to self here: all of the logic outlined in steps 1-5 in the description can instead be implemented using the subgraph, which will be a lot more lightweight.

File: src/Badges.sol [hunk 4]
```diff
@@ -112,21 +129,45 @@ contract Badges is
...
 
-  function createSpecAsRaftHolder(string memory _specUri, uint256 _raftTokenId) external virtual {
-    require(specDataHolder.getRaftOwner(_raftTokenId) == msg.sender, "createSpecAsRaftHolder: unauthorized");
-    require(!specDataHolder.isSpecRegistered(_specUri), "createSpecAsRaftHolder: spec already registered");
-    specDataHolder.setSpecToRaft(_specUri, _raftTokenId);
-    emit SpecCreated(msg.sender, _specUri, _raftTokenId, specDataHolder.getRaftAddress());
-  }
-
-  function createSpecAsHolderOfAnyBadge(
+  function createSpec(
```
Comment (rahulrumalla) [inline src/Badges.sol:132]: Why not have external functions specific to each permission? that way it forces the client to be aware of what permissioned function they're calling. Wdyt?

File: src/Badges.sol [hunk 5]
```diff
@@ -66,14 +72,25 @@ contract Badges is
...
-    // would it be helpful in any way to emit an event here?
-    raftToBadgeSpecCreationPermissions[raftTokenId] = permissionInt;
+  // this is a org-level function that's called from the settings page on a DAO
+  function setSpecCreationPermissions(
+    uint256 _raftTokenId,
+    Permission permissionInt,
+    string[] memory listOfSpecs
+  ) external {
+    // require that the caller is the owner of the raftTokenId
+    require(specDataHolder.getRaftOwner(_raftTokenId) == msg.sender, "setSpecCreationPermissions: unauthorized");
```
Comment (rahulrumalla) [inline src/Badges.sol:82]: could this be a modifier?

File: src/Badges.sol [hunk 6]
```diff
@@ -66,14 +72,25 @@ contract Badges is
...
+
+    // would it be helpful to emit an event here?
```
Comment (rahulrumalla) [inline src/Badges.sol:84]: I think all write actions in general should emit an event so we can index these.

File: src/Badges.sol [hunk 7]
```diff
@@ -66,14 +72,25 @@ contract Badges is
...
+    specCreationPermissions[_raftTokenId] = permissionInt;
+
+    if (permissionInt == Permission.SPECIFIC_BADGE_HOLDERS) {
+      raftToPermittedBadgeSpecCreators[_raftTokenId] = listOfSpecs;
+    }
+  }
+
+  function getBadgesAllowedToCreateSpecs(uint256 _raftTokenId) external view returns (string[] memory) {
```
Comment (rahulrumalla) [inline src/Badges.sol:92]: what's the intention behind this?

File: src/Badges.sol [hunk 8]
```diff
@@ -112,21 +129,45 @@ contract Badges is
...
+    uint256 _badgeId
   ) external virtual {
-    require(owners[badgeId] == msg.sender, "createSpecAsHolderOfAnyBadge: unauthorized");
-    specDataHolder.setSpecToRaft(_specUri, _raftTokenId);
-    emit SpecCreated(msg.sender, _specUri, _raftTokenId, specDataHolder.getRaftAddress());
+    require(!specDataHolder.isSpecRegistered(_specUri), "createSpec: spec already registered");
+
+    Permission specCreationPermission = specCreationPermissions[_raftTokenId];
+
+    if (specCreationPermission == Permission.ONLY_RAFT) {
```
Comment (rahulrumalla) [inline src/Badges.sol:141]: I think this whole thing can be refactored into smaller functions

**Comment by grp06** [comment]
no longer relevant now that we're doing this with spec admins

**Pull request closed without merging**